### PR TITLE
Refactor: Move trace creation into Rule instantiation

### PR DESCRIPTION
**Pull request by kaicataldo**
fixes https://github.com/adobe-community/issues/issues/10947

Moving the trace creation logic into the instantiation of Rules frees us up to extract out a RuleList class, which we want for clarity and to clean up some of the weird things I had to do in my previous  PR.

### CODE REVIEW

**Comment by kaicataldo** [comment]
Let me know if you feel like slowly changing over the stuff I touch to ES6 is adding too much visual  noise to the PR

File: src/Engine.js [hunk 1]
```diff
@@ -259,52 +257,50 @@ Engine = Class.extend({
...
+    arrayDataFromState.forEach((item, index) => {
       // Filtering must happen here instead of a separate filter step to ensure that
       // `index` is consistent between the data from process and the index provided to _insertRule below
       if (filterFunction && !filterFunction(item)) {
         return;
       }
 
-      var toggleSuffix = '';
-      var selectorForItem = descriptor.selector.replace(arrayPropertyRegex, function(match, column) {
+      let toggleSuffix = '';
```
Comment (Attamusc) [inline src/Engine.js]: Does all of this really need to live here, or can this be pushed into the Rule creation? Doesn't seem to be using really anything from here.

**Comment by Attamusc** [comment]
I like the ES6-ification as we go like this

File: src/Engine.js (hunk 1 above)
Comment (kaicataldo) [inline src/Engine.js]: I think it does need to be outside the rule, because the rule has no knowledge of its index in the generated array of rules

Comment (Attamusc) [inline src/Engine.js]: That to me seems weird...why does it _need_ to know it's index? Seems like this is just changing the selector to include the toggleSelector in it with guaranteed uniqueness

File: src/Engine.js [hunk 2]
```diff
@@ -259,52 +257,50 @@ Engine = Class.extend({
...
-    toggleKeys.forEach(function(key) {
-      this.traces[key] = artifactsPrefixed;
-    }, this);
+    return traces;
   },
 
-  _insertRule(selector, spec, arrayMemberExpr, rulesCache) {
-    var rule = new this.constructor.Rule(selector, spec, arrayMemberExpr);
-    var i = this.rules.length;
+  _insertRule(selector, spec, arrayMemberExpr, toggleKeys, togglePrefix, rulesCache) {
```
Comment (hzoo) [inline src/Engine.js]: Might be nice to pass in options for named arguments but just my thought since it can be hard to know what is what `selectorInfo.selector, spec, undefined, selectorInfo.toggleKeys, undefined, rulesContext.rules`

File: src/Engine.js (hunk 1 above)
Comment (kaicataldo) [inline src/Engine.js]: I thought it was to prepend the artifact to denote the index of it in the array of data: https://github.com/behance/focss/pull/44/files#diff-192c40a212fc15f6e5fb1d680297af92R57

File: src/Engine.js (hunk 2 above)
Comment (kaicataldo) [inline src/Engine.js]: Agreed the `undefined` parameters are hard to follow, but not sure I'm following - mind expanding on that idea a bit for me?

File: src/Engine.js (hunk 1 above)
Comment (Attamusc) [inline src/Engine.js]: Ah ok interesting. hmm.

File: src/Engine.js (hunk 2 above)
Comment (Attamusc) [inline src/Engine.js]: Ya, these method signatures are getting long

File: src/Engine.js (hunk 1 above)
Comment (kaicataldo) [inline src/Engine.js]: Maybe @mikesherov has some more insight

File: src/Engine.js (hunk 2 above)
Comment (hzoo) [inline src/Engine.js]: can make some of these an options object (not sure if all or just part of it)

```js
_insertRule({ selector, spec, arrayMemberExpr, toggleKeys, togglePrefix, rulesCache }) {

}
```

Comment (kaicataldo) [inline src/Engine.js]: Gotcha. Thanks for explaining - I like it!

File: src/Engine.js [hunk 3]
```diff
@@ -259,52 +262,57 @@ Engine = Class.extend({
...
-      this.traces[key] = artifactsPrefixed;
-    }, this);
+    return traces;
   },
 
-  _insertRule(selector, spec, arrayMemberExpr, rulesCache) {
-    var rule = new this.constructor.Rule(selector, spec, arrayMemberExpr);
-    var i = this.rules.length;
+  _insertRule(ruleData, { rules }) {
+    const rule = new this.constructor.Rule(ruleData);
```
Comment (kaicataldo) [inline src/Engine.js]: Unrelated to this PR, but was curious what the benefit is of passing in `Rule` as a static property and accessing it through `this.constructor.Rule`. Is it so it's available to the consumer? And if so, why? Doesn't seem like it should be.

File: src/Engine.js (hunk 1 above)
Comment (Attamusc) [inline src/Engine.js]: The more I think about this, the more you're right :P I think this can't live inside the rule because of array rules like `forEach` and `filterEach`. Could be worth investigating as part of the RuleList extraction

Comment (mikesherov) [inline src/Engine.js]: needs to stay. Read the comment *literally* above this discussion.

File: src/Engine.js (hunk 3 above)
Comment (mikesherov) [inline src/Engine.js]: I guess testing somewhere?

Comment (mikesherov) [inline src/Engine.js]: see if changing breaks it.

File: src/Engine.js (hunk 1 above)
Comment (kaicataldo) [inline src/Engine.js]: D'oh :facepalm:. Thanks all

File: src/Engine.js [hunk 4]
```diff
@@ -259,52 +262,57 @@ Engine = Class.extend({
...
+        const newToggleKey = toggleKey + toggleSuffix;
+        selector = selector.replace(`${toggleKey}']?`, `${newToggleKey}']?`);
         return newToggleKey;
       });
 
-      rule = this._insertRule(selectorForItem, descriptor.spec, descriptor.expr + '[' + index + ']', rulesContext.rules);
-      this._addTraces(rule, newToggleKeys, descriptor.expr + '.' + index + '.');
-    }, this);
-  },
+      const ruleData = {
```
Comment (mikesherov) [inline src/Engine.js]: unnecessary temp.

File: src/Engine.js (hunk 3 above)
Comment (kaicataldo) [inline src/Engine.js]: I see, makes sense

File: src/Engine.js (hunk 4 above)
Comment (kaicataldo) [inline src/Engine.js]: Thought it was clearer this way but can def inline it

**Comment by Attamusc** [comment]
👍

**Comment by hzoo** [comment]
👍
